### PR TITLE
onSnapshot support

### PR DESCRIPTION
**Pull request by zeevl**
Hey, first, thanks so much for offering to take over support of this project!

This brings over support for `onSnapshot` that current exists in https://github.com/soumak77/firebase-mock/pull/158.   

I added `metadata` to the `MockFirestoreDocumentSnapshot` so we could use this in our project.  However, I'm not sure what the proper etiquette is for this type of PR... other people's work on the upstream (dead) project, only slightly modified?   They certainly deserve the credit here, haha. 

Anyways, let me know if you want something done differently here.

### CODE REVIEW

**Comment by dmurvihill** [comment]
Hey @zeevl, thanks for the pull request. Your name is on the PR but @BrianChapman's and @Venryx's names are still attached to all of the commits, so I think we should have no trouble with attribution. I'm just returning from a vacation and need a minute to look these over.

**Comment by dmurvihill** [comment]
After merging in master, I am seeing 5 failing unit tests:

```
  463 passing (1s)
  27 pending
  5 failing

  1) MockFirestoreDocument #onSnapshot returns value after document is updated:
     AssertionError: expected undefined to equal 'A new title'
      at test/unit/firestore-document.js:470:41
      at onSnapshot (src/firestore-document.js:9:14768)
      at MockFirestoreDocument.onSnapshot (src/firestore-document.js:9:15093)
      at Context.<anonymous> (test/unit/firestore-document.js:469:11)
      at process.topLevelDomainCallback (domain.js:120:23)

  2) MockFirestoreDocument #onSnapshot returns error if error occured:
     Error: done() called multiple times
      at Suite.<anonymous> (test/unit/firestore-document.js:477:5)
      at Suite.<anonymous> (test/unit/firestore-document.js:467:3)
      at Object.<anonymous> (test/unit/firestore-document.js:14:1)
      at Module._compile (internal/modules/cjs/loader.js:689:30)
      at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
      at Object.Module._extensions.(anonymous function) [as .js] (node_modules/istanbul/lib/hook.js:109:37)
      at Module.load (internal/modules/cjs/loader.js:599:32)
      at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
      at Function.Module._load (internal/modules/cjs/loader.js:530:3)
      at Module.require (internal/modules/cjs/loader.js:637:17)
      at require (internal/modules/cjs/helpers.js:22:18)

  3) MockFirestoreDocument #onSnapshot does not returns value when not updated:

      AssertionError: expected 2 to equal 1
      + expected - actual

      -2
      +1
      
      at Context.<anonymous> (test/unit/firestore-document.js:497:28)
      at process.topLevelDomainCallback (domain.js:120:23)

  4) MockFirestoreDocument #onSnapshot unsubscribes:

      AssertionError: expected 2 to equal 1
      + expected - actual

      -2
      +1
      
      at Context.<anonymous> (test/unit/firestore-document.js:511:28)
      at process.topLevelDomainCallback (domain.js:120:23)

  5) MockFirestoreDocument #onSnapshot accepts option includeMetadataChanges:
     AssertionError: expected undefined to equal 'A new title'
      at test/unit/firestore-document.js:521:41
      at onSnapshot (src/firestore-document.js:9:14768)
      at MockFirestoreDocument.onSnapshot (src/firestore-document.js:9:15093)
      at Context.<anonymous> (test/unit/firestore-document.js:520:11)
      at process.topLevelDomainCallback (domain.js:120:23)



[19:00:00] '<anonymous>' errored after 12 ms
[19:00:00] Error in plugin 'gulp-mocha'
Error
    at Domain.handleException (/Users/dolan/firebase-mock/node_modules/gulp-mocha/index.js:46:24)
    at Domain.emit (events.js:189:13)
    at Domain.EventEmitter.emit (domain.js:441:20)
    at Domain._errorHandler (domain.js:223:23)
    at Object.setUncaughtExceptionCaptureCallback (domain.js:139:29)
    at process._fatalException (internal/bootstrap/node.js:493:31)
[19:00:00] 'test' errored after 417 ms
npm ERR! Test failed.  See above for more details.
```

Does this work on your branch? I'm running a newer version of `node` and `gulp`.

**Comment by zeevl** [comment]
Honestly I haven't been able to get tests to run -- gulp keeps crashing with node.  

What versions are you using?   I've tried with both node 10.x and 12.x and they both crash with node stracktraces, which seem to be incompatibility errors.   Maybe time to upgrade to gulp 4?

**Comment by dmurvihill** [comment]
Gulp 3 is incompatible with Node 10+, but if you merge `dmurvihill:master` you should get Gulp 4.

**Comment by zeevl** [comment]
Perfect, thanks!  Tests fixed.

File: src/firestore-query.js [hunk 1]
```diff
@@ -67,37 +67,11 @@ MockFirestoreQuery.prototype.get = function () {
   var self = this;
   return new Promise(function (resolve, reject) {
     self._defer('get', _.toArray(arguments), function () {
-      var results = {};
+      var results = self._results();
       var limit = 0;
```
Comment (dmurvihill) [inline src/firestore-query.js]: used?

File: src/firestore-query.js [hunk 2]
```diff
@@ -67,37 +67,11 @@ MockFirestoreQuery.prototype.get = function () {
...
-
-            queryable.forEach(function(q) {
-              if (self.limited <= 0 || limit < self.limited) {
-                results[q.key] = _.cloneDeep(q.data);
-                limit++;
-              }
-            });
-          }
-
           resolve(new QuerySnapshot(self.parent === null ? self : self.parent.collection(self.id), results));
```
Comment (dmurvihill) [inline src/firestore-query.js]: I think these two branches are equivalent now.

File: src/firestore-query.js (hunk 1 above)
Comment (zeevl) [reply src/firestore-query.js]: Yes, but `limit` wasn't.   Removed.

File: src/firestore-query.js [hunk 3]
```diff
@@ -169,6 +143,86 @@ MockFirestoreQuery.prototype.limit = function (limit) {
...
+  // onSnapshot should always return when initially called, then
+  // every time data changes.
+  onSnapshot(true);
+  var unsubscribe = this.queue.onPostFlush(onSnapshot);
+
+  // return the unsubscribe function
+  return unsubscribe;
+};
+
+MockFirestoreQuery.prototype._results = function () {
```
[the remark this answers is not in the document]
Comment (zeevl) [reply src/firestore-query.js]: I believe it's setting up the results of any `orderBy` or `limit` query parameters..

File: src/firestore-query.js (hunk 2 above)
Comment (zeevl) [reply src/firestore-query.js]: Not sure what you're referring to here..

**Comment by zeevl** [comment]
Thanks for the quality review @dmurvihill!

File: src/firestore-query.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmurvihill) [reply src/firestore-query.js]: Yeah, on closer inspection it looks like a refactor of the red block in ll. 75-100

File: src/firestore-query.js (hunk 2 above)
Comment (dmurvihill) [reply src/firestore-query.js]: Line 71 and line 73 do the same thing, which means the if-else is meaningless.

**Comment by dmurvihill** [comment]
Merged. Look for a release in the next few days. Thanks @BrianChapman, @Venryx, and @zeevl for getting this over the line.

**Pull request merged**